[rush] Support CI build log folding using env vars

elliot-nelson opened the pull request:
## Summary

Allow monorepo maintainers to fold long build logs in CI using simple header and footer environment variables.

Fixes #2198.

## Details

Since this behavior typically doesn't make sense on local machines, going with environment variables makes the most sense.

Although this behavior is called "grouping" by both GitHub and Azure, I think this word is rather ambiguous; log _folding_ is much more specific and better describes the actual functionality you get. I've proposed `RUSH_BUILD_FOLDING_HEADER` and `RUSH_BUILD_FOLDING_FOOTER` but am open to alternatives!

## How it was tested

 - Tested locally to verify output with env vars on and off, verbose on and off.

iclanton commented on libraries/rush-lib/src/logic/operations/OperationExecutionManager.ts:
```diff
@@ -137,11 +144,23 @@ export class OperationExecutionManager {
 
+      if (!this._quietMode) {
+        if (this._projectLogFooter && this._lastTask) {
+          this._terminal.writeStdoutLine(this._projectLogFooter.replace(/\$\{taskName\}/g, this._lastTask));
```
Extract the token replacement into a function?

iclanton commented on libraries/rush-lib/src/logic/operations/OperationExecutionManager.ts:
```diff
@@ -201,6 +220,12 @@ export class OperationExecutionManager {
 
+    if (!this._quietMode) {
+      if (this._projectLogFooter && this._lastTask) {
+        this._terminal.writeStdoutLine(this._projectLogFooter.replace(/\$\{taskName\}/g, this._lastTask));
```
Seems like there should be a matching header here.

elliot-nelson replied:
In the "stream started" callback for each operation, we write the footer for last section, the separator, and then the header for the next section, and wait until all operations have executed the stream started callback.

At this point, all the operations are done, but we're still "inside" the last folding group that was started, so we need one final line to close it out (if defined).

iclanton replied:
Does this mean that we don't fold the first operation that runs?

It seems like we should write the header, log stuff for the operation, and then write the footer (and it should appear that way in the code). Shouldn't stream-collator be able to handle that?

elliot-nelson replied:
Well, StreamCollator _is_ handling it -- it provides a callback called `onWriterActive` that triggers whenever a new stream starts writing to the collator (including the first stream). We pass this function as that callback, then we just pile all the build log streams into the collator, letting its `onWriterActive` hook write out the fancy header lines as appropriate.

The first stream to write to the collator will trigger this hook, `_lastTask` will be empty so it will skip writing a folding footer, then it will write the header separator, then it will write the folding header.

iclanton commented on libraries/rush-lib/src/api/EnvironmentConfiguration.ts:
```diff
@@ -153,6 +153,44 @@ export enum EnvironmentVariableNames {
+   *   - Travis CI:
+   *     - `travis_fold:start:${taskName}`
+   */
+  RUSH_BUILD_FOLDING_HEADER = 'RUSH_BUILD_FOLDING_HEADER',
```
We've started moving away from the word "build" in favor of "operation" since commands and tasks aren't necessarily builds.

elliot-nelson replied:
:+1: Open to changing this, my reasoning is:

(1) It matches all existing `RUSH_BUILD_` settings like for cache, etc.
(2) Even for people very familiar with Rush, who can parse a paragraph about phases, tasks, and operations, they still think colloquially of "build" (as opposed to update, or install, or check, or publish, etc.). I guess my inclination is to use names like `RUSH_BUILD_` to take advantage of that natural shortcut, and explain in documentation that it applies to any bulk or phased command, rather than being "correct" and calling it `RUSH_OPERATION_`, leaving many casual readers totally adrift.

(Let me know what you think -- I suppose one rebuttal is that the only way to get people used to "operations" is by using it all over, but I don't know that we're really making a concerted effort to do that right now.)

iclanton replied:
I can't stay I have a particularly strong opinion. @octogonz and @dmichon-msft came up with the name IIRC, so they might have some thoughts.

elliot-nelson commented on libraries/rush-lib/src/logic/operations/OperationExecutionManager.ts:
```diff
@@ -137,11 +144,23 @@ export class OperationExecutionManager {
 
+      if (!this._quietMode) {
+        if (this._projectLogFooter && this._lastTask) {
+          this._terminal.writeStdoutLine(this._projectLogFooter.replace(/\$\{taskName\}/g, this._lastTask));
```
The "footer" here is being written before we write the next project separator line (to close out the previous build log section).

elliot-nelson commented:
Incoming changes based on PR feedback:

 - I've changed the `${taskName}` interpolation to be `[taskName]`, to match our existing interpolated values
 - Refactored interpolation to happen in a dedicated method
 - There is now a single env var that controls log folding style (which defaults to off)
 - Your supported log folding styles are configured in `logging.json`

Should be ready for another round.

elliot-nelson commented:
Thanks for that @octogonz!

So I wanted to test out exactly how it looks and feels to collapse all the insides of each operation, and I am worried that perhaps this change is not worth it.

Check out this log output for an operation that is failing:
<img width="677" alt="Screen Shot 2022-12-12 at 10 46 07 PM" src="https://user-images.githubusercontent.com/58273/207221766-f5900645-b387-4de5-b191-2a3586a49960.png">

I can see that this operation is failing in the bottom of the build log, with its summary output; but if I scroll up to find it and view the full log, I basically have to search by operation name, because nothing about the output indicates a problem.

🤔 I'm not sure what improvement would be required. Perhaps if the log is folded, and the task ended in warning or error, an additional line _below_ the fold could print the result of the operation so it's visible when quickly scanning the log output?
